File: server/webapp/WEB-INF/rails.new/app/models/api_v1/go_latest_version.rb
##########################################################################
# Copyright 2015 ThoughtWorks, Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
##########################################################################

module ApiV1
  class GoLatestVersion
    def initialize(latest_version, system_environment)
      @message = latest_version[:message]
      @message_signature = latest_version[:message_signature]
      @signing_public_key = latest_version[:signing_public_key]
      @signing_public_key_signature = latest_version[:signing_public_key_signature]
      @system_environment = system_environment
    end

    def valid?
      !signing_public_key_tampered? and !message_tampered?
    end

    def latest_version
      JSON.parse(@message)['latest-version']
    end

    private

    def message_tampered?
      !MessageVerifier.verify(@message, @message_signature, @signing_public_key)
    end

    def signing_public_key_tampered?
      public_key = File.read(@system_environment.getUpdateServerPublicKeyPath())

      !MessageVerifier.verify(@signing_public_key, @signing_public_key_signature, public_key)
    end
  end
end

File: api/api-version-infos-v1/src/main/java/com/thoughtworks/go/apiv1/versioninfos/models/GoLatestVersion.java
/*
 * Copyright Thoughtworks, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.thoughtworks.go.apiv1.versioninfos.models;

import com.thoughtworks.go.util.SystemEnvironment;
import com.thoughtworks.go.util.json.JsonHelper;
import lombok.Data;
import lombok.experimental.Accessors;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.security.InvalidKeyException;
import java.security.NoSuchAlgorithmException;
import java.security.SignatureException;
import java.security.spec.InvalidKeySpecException;
import java.util.Map;

import static java.nio.charset.StandardCharsets.UTF_8;

@Data
@Accessors(chain = true)
public class GoLatestVersion {
    private static final Logger LOG = LoggerFactory.getLogger(GoLatestVersion.class);
    private SystemEnvironment systemEnvironment;
    private String message;
    private String messageSignature;
    private String signingPublicKey;
    private String signingPublicKeySignature;

    public String latestVersion() {
        return (String) JsonHelper.fromJson(message, Map.class).get("latest-version");
    }

    public boolean isValid() {
        try {
            return !signingPublicKeyTampered() && !messageTampered();
        } catch (Exception e) {
            LOG.error("Exception occurred while verifying message and public key.", e);
            return false;
        }
    }

    private boolean signingPublicKeyTampered() throws IOException, NoSuchAlgorithmException, InvalidKeySpecException, InvalidKeyException, SignatureException {
        String publicKey = Files.readString(Path.of(systemEnvironment.getUpdateServerPublicKeyPath()), UTF_8);
        return !EncryptionHelper.verifyRSASignature(signingPublicKey, signingPublicKeySignature, publicKey);
    }

    private boolean messageTampered() throws NoSuchAlgorithmException, SignatureException, InvalidKeyException, InvalidKeySpecException {
        return !EncryptionHelper.verifyRSASignature(message, messageSignature, signingPublicKey);
    }
}
